{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/sm/rgfxp6g94f18340p2rvqnsc00000gn/T/ipykernel_9623/2657062766.py:1: DtypeWarning: Columns (12,13,14,40,46,55,65,66,67,68,177,294,298,299,300,301,304,305,306,307,308,309,310,318,320,328,330,346,348,350,352,354,359,361,363,366,373,377,381,383,418,419,432,433,434,435,436,438,439,440,441,442,444,445,446,447,448,450,451,452,453,454) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  metadata = pd.read_csv('/Volumes/PGH-Backup/ibd_data/metadata/hmp2_metadata_2018-08-20.csv')\n"
     ]
    }
   ],
   "source": [
    "metadata = pd.read_csv('/Volumes/PGH-Backup/ibd_data/metadata/hmp2_metadata_2018-08-20.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnaseq = metadata[metadata['data_type'] == 'host_transcriptomics']\n",
    "\n",
    "# display(rnaseq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "metagenomics = pd.read_csv('/Volumes/PGH-Backup/ibd_data/metadata/hmp2_metagenomics_metadata.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Project</th>\n",
       "      <th>External ID</th>\n",
       "      <th>Participant ID</th>\n",
       "      <th>site_sub_coll</th>\n",
       "      <th>data_type</th>\n",
       "      <th>week_num</th>\n",
       "      <th>date_of_receipt</th>\n",
       "      <th>interval_days</th>\n",
       "      <th>visit_num</th>\n",
       "      <th>Research Project</th>\n",
       "      <th>...</th>\n",
       "      <th>FecalCal received at MGH:</th>\n",
       "      <th>Proteomics received at LBNL:</th>\n",
       "      <th>Stool Sample ID:  Tube A  (EtOH)</th>\n",
       "      <th>Sample ID: Tube B (No preservative)</th>\n",
       "      <th>Tube A and B received at Broad:</th>\n",
       "      <th>stool_id</th>\n",
       "      <th>smoking status</th>\n",
       "      <th>Number years smoked</th>\n",
       "      <th>Age when started smoking</th>\n",
       "      <th>How many cigarettes/cigars/etc. do you smoke per day?</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>G79889</td>\n",
       "      <td>CSM5FZ3N_P</td>\n",
       "      <td>C3001</td>\n",
       "      <td>C3001C1</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2014-03-14</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-5FZ3N</td>\n",
       "      <td>SM-5FZ3P</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G79894</td>\n",
       "      <td>CSM5FZ3R_P</td>\n",
       "      <td>C3001</td>\n",
       "      <td>C3001C2</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2014-03-28</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-5FZ3R</td>\n",
       "      <td>SM-5UJBB</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G79903</td>\n",
       "      <td>CSM5YRY7_P</td>\n",
       "      <td>C3001</td>\n",
       "      <td>C3001C3</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>4.0</td>\n",
       "      <td>2014-04-15</td>\n",
       "      <td>18.0</td>\n",
       "      <td>6</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-5YRY7</td>\n",
       "      <td>SM-5YRY8</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>G79913</td>\n",
       "      <td>CSM5FZ3V_P</td>\n",
       "      <td>C3001</td>\n",
       "      <td>C3001C4</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2014-04-28</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-5FZ3V</td>\n",
       "      <td>SM-5FZ3W</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G79926</td>\n",
       "      <td>CSM5FZ4C_P</td>\n",
       "      <td>C3001</td>\n",
       "      <td>C3001C5</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2014-05-09</td>\n",
       "      <td>11.0</td>\n",
       "      <td>8</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-5FZ4C</td>\n",
       "      <td>SM-5FZ4D</td>\n",
       "      <td>Yes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1633</th>\n",
       "      <td>G111931</td>\n",
       "      <td>PSMA26A1</td>\n",
       "      <td>P6038</td>\n",
       "      <td>P6038C12</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>24.0</td>\n",
       "      <td>2016-04-28</td>\n",
       "      <td>14.0</td>\n",
       "      <td>16</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-A26A1</td>\n",
       "      <td>SM-A26A2</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1634</th>\n",
       "      <td>G110660</td>\n",
       "      <td>PSMA26A3</td>\n",
       "      <td>P6038</td>\n",
       "      <td>P6038C13</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>26.0</td>\n",
       "      <td>2016-05-12</td>\n",
       "      <td>14.0</td>\n",
       "      <td>18</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-A26A3</td>\n",
       "      <td>SM-A26A4</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1635</th>\n",
       "      <td>G111932</td>\n",
       "      <td>PSMB4MC1</td>\n",
       "      <td>P6038</td>\n",
       "      <td>P6038C17</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>34.0</td>\n",
       "      <td>2016-07-08</td>\n",
       "      <td>8.0</td>\n",
       "      <td>22</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-B4MC1</td>\n",
       "      <td>SM-B4MC2</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1636</th>\n",
       "      <td>G111934</td>\n",
       "      <td>PSMB4MC3</td>\n",
       "      <td>P6038</td>\n",
       "      <td>P6038C18</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>35.0</td>\n",
       "      <td>2016-07-20</td>\n",
       "      <td>12.0</td>\n",
       "      <td>23</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-B4MC3</td>\n",
       "      <td>SM-B4MC4</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1637</th>\n",
       "      <td>G110704</td>\n",
       "      <td>PSMB4MC5</td>\n",
       "      <td>P6038</td>\n",
       "      <td>P6038C19</td>\n",
       "      <td>metagenomics</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2016-08-03</td>\n",
       "      <td>14.0</td>\n",
       "      <td>25</td>\n",
       "      <td>ibdmdb</td>\n",
       "      <td>...</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>SM-B4MC5</td>\n",
       "      <td>SM-B4MC6</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1638 rows × 490 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Project External ID Participant ID site_sub_coll     data_type  \\\n",
       "0      G79889  CSM5FZ3N_P          C3001       C3001C1  metagenomics   \n",
       "1      G79894  CSM5FZ3R_P          C3001       C3001C2  metagenomics   \n",
       "2      G79903  CSM5YRY7_P          C3001       C3001C3  metagenomics   \n",
       "3      G79913  CSM5FZ3V_P          C3001       C3001C4  metagenomics   \n",
       "4      G79926  CSM5FZ4C_P          C3001       C3001C5  metagenomics   \n",
       "...       ...         ...            ...           ...           ...   \n",
       "1633  G111931    PSMA26A1          P6038      P6038C12  metagenomics   \n",
       "1634  G110660    PSMA26A3          P6038      P6038C13  metagenomics   \n",
       "1635  G111932    PSMB4MC1          P6038      P6038C17  metagenomics   \n",
       "1636  G111934    PSMB4MC3          P6038      P6038C18  metagenomics   \n",
       "1637  G110704    PSMB4MC5          P6038      P6038C19  metagenomics   \n",
       "\n",
       "      week_num date_of_receipt  interval_days  visit_num Research Project  \\\n",
       "0          0.0      2014-03-14            0.0          4           ibdmdb   \n",
       "1          2.0      2014-03-28           14.0          5           ibdmdb   \n",
       "2          4.0      2014-04-15           18.0          6           ibdmdb   \n",
       "3          6.0      2014-04-28           13.0          7           ibdmdb   \n",
       "4          8.0      2014-05-09           11.0          8           ibdmdb   \n",
       "...        ...             ...            ...        ...              ...   \n",
       "1633      24.0      2016-04-28           14.0         16           ibdmdb   \n",
       "1634      26.0      2016-05-12           14.0         18           ibdmdb   \n",
       "1635      34.0      2016-07-08            8.0         22           ibdmdb   \n",
       "1636      35.0      2016-07-20           12.0         23           ibdmdb   \n",
       "1637      37.0      2016-08-03           14.0         25           ibdmdb   \n",
       "\n",
       "      ... FecalCal received at MGH:  Proteomics received at LBNL:  \\\n",
       "0     ...                        No                            No   \n",
       "1     ...                        No                            No   \n",
       "2     ...                        No                            No   \n",
       "3     ...                        No                            No   \n",
       "4     ...                        No                            No   \n",
       "...   ...                       ...                           ...   \n",
       "1633  ...                        No                            No   \n",
       "1634  ...                        No                            No   \n",
       "1635  ...                        No                            No   \n",
       "1636  ...                        No                            No   \n",
       "1637  ...                        No                            No   \n",
       "\n",
       "     Stool Sample ID:  Tube A  (EtOH) Sample ID: Tube B (No preservative)  \\\n",
       "0                            SM-5FZ3N                            SM-5FZ3P   \n",
       "1                            SM-5FZ3R                            SM-5UJBB   \n",
       "2                            SM-5YRY7                            SM-5YRY8   \n",
       "3                            SM-5FZ3V                            SM-5FZ3W   \n",
       "4                            SM-5FZ4C                            SM-5FZ4D   \n",
       "...                               ...                                 ...   \n",
       "1633                         SM-A26A1                            SM-A26A2   \n",
       "1634                         SM-A26A3                            SM-A26A4   \n",
       "1635                         SM-B4MC1                            SM-B4MC2   \n",
       "1636                         SM-B4MC3                            SM-B4MC4   \n",
       "1637                         SM-B4MC5                            SM-B4MC6   \n",
       "\n",
       "     Tube A and B received at Broad:  stool_id  smoking status  \\\n",
       "0                                 No       NaN             NaN   \n",
       "1                                Yes       NaN             NaN   \n",
       "2                                Yes       NaN             NaN   \n",
       "3                                Yes       NaN             NaN   \n",
       "4                                Yes       NaN             NaN   \n",
       "...                              ...       ...             ...   \n",
       "1633                              No       NaN             NaN   \n",
       "1634                              No       NaN             NaN   \n",
       "1635                              No       NaN             NaN   \n",
       "1636                              No       NaN             NaN   \n",
       "1637                              No       NaN             NaN   \n",
       "\n",
       "      Number years smoked  Age when started smoking  \\\n",
       "0                     NaN                       NaN   \n",
       "1                     NaN                       NaN   \n",
       "2                     NaN                       NaN   \n",
       "3                     NaN                       NaN   \n",
       "4                     NaN                       NaN   \n",
       "...                   ...                       ...   \n",
       "1633                  NaN                       NaN   \n",
       "1634                  NaN                       NaN   \n",
       "1635                  NaN                       NaN   \n",
       "1636                  NaN                       NaN   \n",
       "1637                  NaN                       NaN   \n",
       "\n",
       "      How many cigarettes/cigars/etc. do you smoke per day?  \n",
       "0                                                   NaN      \n",
       "1                                                   NaN      \n",
       "2                                                   NaN      \n",
       "3                                                   NaN      \n",
       "4                                                   NaN      \n",
       "...                                                 ...      \n",
       "1633                                                NaN      \n",
       "1634                                                NaN      \n",
       "1635                                                NaN      \n",
       "1636                                                NaN      \n",
       "1637                                                NaN      \n",
       "\n",
       "[1638 rows x 490 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(metagenomics)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter df1 based on the Participant ID column in df2\n",
    "metagenomics_filt = metagenomics[metagenomics['Participant ID'].isin(rnaseq['Participant ID'])]\n",
    "\n",
    "# print(metagenomics_filt['Participant ID'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Participant ID</th>\n",
       "      <th>week_num</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>C3002</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>C3002</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>272</th>\n",
       "      <td>C3002</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>273</th>\n",
       "      <td>C3002</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>274</th>\n",
       "      <td>C3003</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>P6024</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>P6025</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>P6025</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>520</th>\n",
       "      <td>P6028</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>521</th>\n",
       "      <td>P6028</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>252 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    Participant ID  week_num\n",
       "270          C3002       0.0\n",
       "271          C3002       0.0\n",
       "272          C3002       0.0\n",
       "273          C3002       0.0\n",
       "274          C3003       1.0\n",
       "..             ...       ...\n",
       "517          P6024       2.0\n",
       "518          P6025       0.0\n",
       "519          P6025       0.0\n",
       "520          P6028       2.0\n",
       "521          P6028       2.0\n",
       "\n",
       "[252 rows x 2 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rna_subset = rnaseq[['Participant ID', 'week_num']]\n",
    "display(rna_subset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_group = rna_subset.groupby('Participant ID').max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_group = pd.DataFrame(rna_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>week_num</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Participant ID</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>C3002</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3003</th>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3004</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3005</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C3006</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P6017</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P6018</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P6024</th>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P6025</th>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>P6028</th>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>90 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                week_num\n",
       "Participant ID          \n",
       "C3002                0.0\n",
       "C3003                1.0\n",
       "C3004                0.0\n",
       "C3005                0.0\n",
       "C3006                0.0\n",
       "...                  ...\n",
       "P6017                0.0\n",
       "P6018                0.0\n",
       "P6024                2.0\n",
       "P6025                0.0\n",
       "P6028                2.0\n",
       "\n",
       "[90 rows x 1 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(rna_group)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_closest_week(participant_id, week_num, metagenomic_df):\n",
    "    # Filter the metagenomic_df for the same Participant ID\n",
    "    participant_metagenomic = metagenomic_df[metagenomic_df[\"Participant ID\"] == participant_id]\n",
    "    \n",
    "    # If no metagenomic samples for this participant, return None\n",
    "    if participant_metagenomic.empty:\n",
    "        return None\n",
    "    \n",
    "    # Calculate the difference between RNAseq week_num and metagenomic week_num\n",
    "    participant_metagenomic[\"diff\"] = participant_metagenomic[\"week_num\"] - week_num\n",
    "\n",
    "    print(participant_metagenomic['Participan ID', 'diff'])\n",
    "\n",
    "    # # Select the sample with the smallest non-negative difference (closest week >= RNAseq week)\n",
    "    # after_or_same = participant_metagenomic[participant_metagenomic[\"diff\"] >= 0].sort_values(by=\"diff\")\n",
    "    # if not after_or_same.empty:\n",
    "    #     return after_or_same.iloc[0]\n",
    "    \n",
    "    # # If no samples are found, select the sample with the smallest negative difference (closest week < RNAseq week)\n",
    "    # before = participant_metagenomic[participant_metagenomic[\"diff\"] < 0].sort_values(by=\"diff\", ascending=False)\n",
    "    # if not before.empty:\n",
    "    #     return before.iloc[0]\n",
    "    \n",
    "    # # If no match is found, return None\n",
    "    # return None\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for index, row in rna_group.iterrows():\n",
    "    print(row['week_num'])\n",
    "\n",
    "    find_closest_week(index, row['week_num'], metagenomics_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize an empty list to store the matched rows\n",
    "matched_rows = []\n",
    "\n",
    "# Iterate over each row in the RNAseq DataFrame\n",
    "for index, row in rna_group.iterrows():\n",
    "    participant_id = row.name  # Assuming 'Participant ID' is the index\n",
    "    week_num = row['week_num']\n",
    "    \n",
    "    # Find the closest metagenomic sample for this participant\n",
    "    matched_row = find_closest_week(participant_id, week_num, metagenomics_filt)\n",
    "    \n",
    "    # If a match is found, append it to the list of matched rows\n",
    "    if matched_row is not None:\n",
    "        matched_rows.append(matched_row)\n",
    "\n",
    "# Convert the list of matched rows into a new DataFrame\n",
    "matched_df = pd.DataFrame(matched_rows)\n",
    "\n",
    "# Merge the RNAseq DataFrame with the matched metagenomic samples\n",
    "merged_df = rna_group.merge(matched_df, on=\"Participant ID\", suffixes=('_rna', '_metagenomic'))\n",
    "\n",
    "# Print or use the resulting DataFrame\n",
    "display(merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_df = pd.merge(rnaseq, metagenomics_filt, on='Participant ID', how='left', suffixes=('_rnaseq', '_metagenomics'))\n",
    "\n",
    "display(merged_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "viewing = merged_df[['Participant ID', 'week_num_rnaseq', 'week_num_metagenomics', 'Project_metagenomics']]\n",
    "\n",
    "display(viewing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter for the closest timepoint or earliest if none is before or at the RNA-seq timepoint\n",
    "def find_closest_timepoint(group):\n",
    "    # Filter for metagenomic samples that are before or at the RNA-seq timepoint\n",
    "    valid_samples = group[group['week_num_metagenomics'] <= group['week_num_rnaseq'].iloc[0]]\n",
    "    \n",
    "    if not valid_samples.empty:\n",
    "        # Find the sample with the closest (maximum) timepoint\n",
    "        closest_sample = valid_samples.loc[valid_samples['week_num_metagenomics'].idxmax()]\n",
    "    else:\n",
    "        # If no valid sample exists, return the earliest metagenomic sample\n",
    "        closest_sample = group.loc[group['week_num_metagenomics'].idxmin()]\n",
    "    \n",
    "    return pd.Series([closest_sample['External ID_metagenomics'], closest_sample['week_num_metagenomics']], \n",
    "                     index=['External ID_metagenomics', 'week_num_metagenomics'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "closest_samples = merged_df.groupby(['Participant ID', 'Project_rnaseq']).apply(find_closest_timepoint).reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnaseq = rnaseq.rename(columns={'Project': 'Project_rnaseq'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge the closest samples back to the original RNA-seq DataFrame\n",
    "final_df = pd.merge(rnaseq, closest_samples, on=['Participant ID', 'Project_rnaseq'], how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df['External ID_metagenomics']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_df.to_csv('/Volumes/PGH-Backup/ibd_data/metadata/matched_rnaseq_mgx_sample_mapping.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for_viewing = final_df[['Participant ID', 'External ID', 'External ID_metagenomics', 'week_num', 'week_num_metagenomics', 'Project_rnaseq']]\n",
    "\n",
    "display(for_viewing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for_viewing.to_csv('/Volumes/PGH-Backup/ibd_data/metadata/matched_rnaseq_mgx_sample_mapping.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Use Mapping File to Join MGX and RNA Data for Appropriate Samples**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load datasets\n",
    "mgx = pd.read_csv('/Volumes/PGH-Backup/ibd_data/humann_second_run/ibd_genefamilies_relab_clustered.tsv', sep='\\t')\n",
    "rna = pd.read_csv('/Volumes/PGH-Backup/ibd_data/rnaseq/tmm_normalized_counts.tsv', sep='\\t')\n",
    "mapping = pd.read_csv('/Volumes/PGH-Backup/ibd_data/metadata/matched_rnaseq_mgx_sample_mapping.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Search for a value in a specific column and return the corresponding row\n",
    "row = mgx.loc[mgx['sample_id'] == 'CSM5FZ3T']\n",
    "\n",
    "# Display the row\n",
    "print(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tanspose so that Genes are columns and samples are rows in RNAseq data\n",
    "rna = rna.set_index('Gene').T\n",
    "\n",
    "rna.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(mgx)\n",
    "display(rna)\n",
    "display(mapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna = rna.reset_index().rename(columns={'index': 'sample_id'})\n",
    "display(rna.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create list of enriched NOD2 genes\n",
    "with open ('/Volumes/PGH-Backup/ibd_data/rnaseq/GSE22611_NOD2_VS_CTRL_TRANSDUCED_HEK293T_CELL_UP.v2024.1.Hs.grp', 'r') as f:\n",
    "    gene_list = [line.strip() for line in f.readlines()]\n",
    "    gene_list = gene_list[2:]\n",
    "\n",
    "print(gene_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset for genes present in table\n",
    "present_genes = [gene for gene in gene_list if gene in rna.columns]\n",
    "\n",
    "present_genes.append('sample_id')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Subset rnaseq data\n",
    "rna_filt = rna[present_genes]\n",
    "\n",
    "display(rna_filt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # write out test subset rnaseq file:\n",
    "# rna_filt.to_csv('/Volumes/PGH-Backup/ibd_data/rnaseq/test_subset_rnaseq.tsv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_mgx = pd.merge(rna_filt, mapping, left_on='sample_id', right_on='External ID', how='inner')\n",
    "\n",
    "display(merged_mgx)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_mgx['External ID_metagenomics'] = merged_mgx['External ID_metagenomics'].str.replace('_P', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_mgx_rna = pd.merge(merged_mgx, mgx, left_on='External ID_metagenomics', right_on= 'sample_id', how='left')\n",
    "\n",
    "display(merged_mgx_rna)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "subset_df = merged_mgx_rna[['sample_id_x', 'sample_id_y', 'External ID_metagenomics']]\n",
    "\n",
    "display(subset_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_mgx_rna.to_csv('/Volumes/PGH-Backup/ibd_data/rnaseq/test_subset_rnaseq_mgx.tsv', index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "view = pd.read_csv('/Volumes/PGH-Backup/ibd_data/rnaseq/test_subset_rnaseq_mgx.tsv', sep='\\t')\n",
    "\n",
    "display(view)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
